### ISSUE TITLE
CRT library uses massive CPU on large instances

### ISSUE THREAD

**Issue by LarryFinn** [awslabs/aws-crt-java#836]
### Describe the bug

We recently upgraded our crt library from 0.24.0 to 0.31.1 and noticed massive cpu spikes.  It seems that the library tries to use all the cpu the machine has, which is beyond what the kube container is allocated.  I've tried toggling `maxConcurrency` and `targetThroughputInGbps` but it has no affects.  I've decreased the "apparent" cpu count in java by using -XX:ActiveProcessorCount=10.  The very odd part is in datadog the cpu spikes seem to be from lock functions.  Attached are two files during a load test.  You can see the change in cpu time for the underlying library

![Screenshot 2024-10-11 at 9 35 05 AM](https://github.com/user-attachments/assets/303f6e9a-f036-4b2e-aa35-83566f440176)
![Screenshot 2024-10-11 at 9 33 34 AM](https://github.com/user-attachments/assets/9367274c-f93d-4c4f-b2e5-4dda6d1d6eca)

### Expected Behavior

Should not utilize all cpu available

### Current Behavior

Utilizes all cpu available

### Reproduction Steps

We see this in load tests, i am not sure how to reproduce it reliably

### aws-crt-java version used

0.31.1

### Java version used

17

### Operating System and version

Ubuntu 20.04.6 x86_64

**Comment by jmklix** [comment]
Can you provide a minimal reproduction code sample that produces the same CPU spikes that you're seeing? Does this only happen when you're running on a large instance?

**Comment by LarryFinn** [comment]
Sorry it took so long, work got busy and it was hard to make this minimal.  I think there are two things at play here:
1. we use multiple s3 clients, which isn't a best practice but i wouldnt expect this outcome
2. we upload as we download
The problem is definitely more related to #1, but im surprised it happens.  Here's some code, it might not compile exactly because i had some different inputs but i think you can figure it out
```
var s3CrtAsyncClientBuilder =  S3AsyncClient.crtBuilder()
                .maxConcurrency(Integer.MAX_VALUE);
   private void multiClientUpload() {
        var concurrency = 30;
        var futures = new ArrayList<CompletableFuture<Integer>>(concurrency);
        var clients = new ArrayList<S3AsyncClient>(concurrency * 2);
        var managers = new ArrayList<S3TransferManager>(concurrency * 2);
        long fileSize = fileSize(SOME_PATH);
        for (int i = 0; i < concurrency; i++) {
            var dlClient = s3CrtAsyncClientBuilder.build();
            clients.add(dlClient);
            var dlMgr = S3TransferManager.builder().s3Client(dlClient).build();
            managers.add(dlMgr);
            var ulClient = s3CrtAsyncClientBuilder.build();
            clients.add(ulClient);
            var ulMgr = S3TransferManager.builder().s3Client(ulClient).build();
            managers.add(ulMgr);
            futures.add(doRequestAndUpload(SOME_PATH, i, dlMgr, ulMgr, fileSize));
        }
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
        managers.forEach(S3TransferManager::close);
        clients.forEach(S3AsyncClient::close);
    }

    private CompletableFuture<Integer> doRequestAndUpload(
            String path
            int i,
            S3TransferManager mgr,
            S3TransferManager uploadMgr,
            long fileSize) {
        return doRequest(path, i, mgr, stream -> {
            var body =
                    new InputStreamWithExecutorAsyncRequestBody(AsyncRequestBodyFromInputStreamConfiguration.builder()
                            .executor(testPool)
                            .inputStream(stream)
                            .maxReadLimit(8 * 1024 * 1024)
                            .contentLength(fileSize)
                            .build());
            var uploadRequest = UploadRequest.builder()
                    .requestBody(body)
                    .putObjectRequest(PutObjectRequest.builder()
                            .bucket(BUCKET_HERE)
                            .key("blarg/" + i + ".csv")
                            .build())
                    .build();
            var result = uploadMgr.upload(uploadRequest).completionFuture().join();
            return i;
        });
    }

    private long fileSize(String path) {
        try (var client = s3CrtAsyncClientBuilder.build()) {
            return client.headObject(b -> b.bucket(BUCKET_HERE).key(path))
                    .thenApply(HeadObjectResponse::contentLength)
                    .join();
        }
    }
    private CompletableFuture<Integer> doRequest(
            String path,
            int i,
            S3TransferManager mgr,
            Function<InputStream, Integer> consumer) {
        DownloadRequest<ResponseInputStream<GetObjectResponse>> downloadRequest;
        var downloadRequestBuilder = DownloadRequest.builder()
                .getObjectRequest(GetObjectRequest.builder()
                        .bucket(SOME_BUCKET)
                        .key(path)
                        .build());
            downloadRequest = downloadRequestBuilder
                    .responseTransformer(AsyncResponseTransformer.toBlockingInputStream())
                    .build();
        return mgr.download(downloadRequest)
                .completionFuture()
                .thenApplyAsync(
                        input -> {
                            return consumer.apply(input.result());
                        },
                        testPool);
    }

```
`multiClientUpload` is what triggers the whole thing.  I made a cached threadpool as testPool.  The file i used is about 500mb with ~6m lines.  my guess is the multiple clients need to do some locking JNI wise and that locks across clients

**Comment by waahm7** [comment]
@LarryFinn Thank you for the code. Can you please explain a bit more about the exact problem that you are facing? Is it causing the application to crash, or is it slow, etc.?
> noticed massive cpu spikes. It seems that the library tries to use all the cpu the machine has, which is beyond what the kube container is allocated. I've tried toggling maxConcurrency and targetThroughputInGbps but it has no affects. 

Each client of CRT will create N threads, but they are green threads and will not have much overhead. Number of threads doesn't correlate with the concurrency. You can limit the overall concurrency by lowering maxConcurrency and targetThroughput. So, even though CRT will create N threads, it should not have much overhead as we will be limiting the concurrency per client.

> we use multiple s3 clients, which isn't a best practice but i wouldnt expect this outcome

Any reason you need to do multiple S3 clients? It is not a best practice since each client will create it's own thread pool and other resources and will lead to duplicated work and resources.

**Comment by LarryFinn** [comment]
It seems like the clients are waiting to acquire JNI locks.  Thats all i can discern from the profiling information.  When I don't use multiple clients, our code actually get stuck a bit.  even if using multiple clients isnt a best practice, i wouldnt expect the cpu jump.  
I ran the snipped above with shared clients vs not shared clients, and the cpu usage is like 10x with multiple clients, which is really bizarre

**Comment by waahm7** [comment]
> When I don't use multiple clients, our code actually get stuck a bit. 
Can you explain a bit more about the use case and the error that you are getting? I think it might be more useful to focus on getting one client to work properly.

> Even if using multiple clients isnt a best practice, i wouldnt expect the cpu jump.
That will depend on how many clients are getting created in parallel. Each client will have its own thread pool and will use the CPU. So N clients will have N times the resource usage.

**Comment by LarryFinn** [comment]
> Even if using multiple clients isnt a best practice, i wouldnt expect the cpu jump

Im not an expert on the C code, but it def looks like the datadog profiler is showing massive amount of time and cpu spent in locking, which doesnt make sense, even with more threads doing more things

> Can you explain a bit more about the use case, workload and the error that you are getting? I think it might be more useful to focus on getting one client to work properly.

Yeah, we store a lot of results in s3 and what we will do is some set operations on multiple s3 files.  so as we download s3 files we process them, and upload the results.  it's hard to say what happens with a single shared client but sometimes it just gets stuck.  java thread dumps dont really show anything too interesting, and im not sure if there is a better way to debug what's going on

**Comment by LarryFinn** [comment]
@waahm7 i created a simple example to show issues we've been having.  it's contrived but whatever
This method with sharing a single client gets stuck
```
   public void manual() throws IOException {
        var client = s3AsyncClientBuilder.build();
        var mgr = S3TransferManager.builder().s3Client(client).build();
        var file = Path.of("/tmp/some-long-csv.csv");
        var length = file.toFile().length();
        var streams = new ArrayList<PipedOutputStream>();
        var futures = new ArrayList<CompletableFuture<CompletedUpload>>();
        for (int i =0; i < 10; i++) {
            var putRequest = PutObjectRequest.builder().bucket(my-bucket).key("prefix/" + i).build();
            PipedOutputStream outputStream = new PipedOutputStream();
            streams.add(outputStream);
            InputStream inputStream = new PipedInputStream(outputStream);

            var uploadRequest = UploadRequest.builder()
                .putObjectRequest(putRequest)
                .requestBody(AsyncRequestBody.fromInputStream(inputStream, length, uploadExecutor))
                .build();

            futures.add(mgr.upload(uploadRequest).completionFuture());
        }
        try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
            String line;
            while ((line = fis.readLine()) != null) {
                for (var stream : streams) {
                    stream.write((line + "\n").getBytes());
                }
            }
        }
        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
    }
```

but this one doesnt
```
  public void manualNotShared() throws IOException {
        var file = Path.of("/tmp/some-long-csv.csv");
        var length = file.toFile().length();
        var streams = new ArrayList<PipedOutputStream>();
        var futures = new ArrayList<CompletableFuture<CompletedUpload>>();
        for (int i =0; i < 10; i++) {
            var client = s3AsyncClientBuilder.build();
            var mgr = S3TransferManager.builder().s3Client(client).build();
            var putRequest = PutObjectRequest.builder().bucket(my-bucket).key("prefix/" + i).build();
            PipedOutputStream outputStream = new PipedOutputStream();
            streams.add(outputStream);
            InputStream inputStream = new PipedInputStream(outputStream);

            var uploadRequest = UploadRequest.builder()
                .putObjectRequest(putRequest)
                .requestBody(AsyncRequestBody.fromInputStream(inputStream, length, uploadExecutor))
                .build();

            futures.add(mgr.upload(uploadRequest).completionFuture());
        }
        try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
            String line;
            while ((line = fis.readLine()) != null) {
                for (var stream : streams) {
                    stream.write((line + "\n").getBytes());
                }
            }
        }
        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
    }
```

**Comment by TingDaoK** [comment]
Let's focus on the single client result in hanging first!

So, I tried your code sample with minor change to add the executors
```
    private static final Logger logger = LoggerFactory.getLogger(App.class);

    public static void main(String... args) throws Exception {
        S3AsyncClient client = S3AsyncClient.builder().build();
        S3TransferManager mgr = S3TransferManager.builder().s3Client(client).build();
        Path file = Path.of("/Users/dengket/project/graalvm/test-crt/sample-project-crt/file.txt");
        long length = file.toFile().length();
        ArrayList<PipedOutputStream> streams = new ArrayList<PipedOutputStream>();
        ArrayList<InputStream> input_streams = new ArrayList<InputStream>();
        var futures = new ArrayList<CompletableFuture<CompletedUpload>>();

        // Executor service for asynchronous operations
        ExecutorService uploadExecutor = Executors.newFixedThreadPool(20);  // Define thread pool for upload tasks

        for (int i =0; i < 20; i++) {
            var putRequest = PutObjectRequest.builder().bucket(bucket).key("prefix/" + i).build();
            PipedOutputStream outputStream = new PipedOutputStream();
            streams.add(outputStream);
            InputStream inputStream = new PipedInputStream(outputStream);
            input_streams.add(inputStream);

            var uploadRequest = UploadRequest.builder()
                    .putObjectRequest(putRequest)
                    .requestBody(AsyncRequestBody.fromInputStream(inputStream, length, uploadExecutor))
                    .build();

            futures.add(mgr.upload(uploadRequest).completionFuture());
        }

        // Run the file reading and writing task asynchronously
        uploadExecutor.submit(() -> {
            try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
                String line;
                while ((line = fis.readLine()) != null) {
                    for (var stream : streams) {
                        stream.write((line + "\n").getBytes());
                    }
                }

                // Close all PipedOutputStreams after writing is done
                for (var stream : streams) {
                    stream.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        });

        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
        mgr.close();
        client.close();
        uploadExecutor.shutdown();
    }

```

And I cannot reproduce the hanging. 
I used a 1 MB file to test, it works fine, but when I bump the file size, I got bunch of errors seems to because of the connection been idle too long to send any data. And, it's hanging, I got bunch of failures reported as expected.

**Comment by LarryFinn** [comment]
@TingDaoK you didn't use the CRT client in your example.  
you have `S3AsyncClient client = S3AsyncClient.builder().build();` but it should be 
`S3AsyncClient client = S3AsyncClient.crtBuilder().build()`
i believe your example will fail if you use crtBuilder too

**Comment by LarryFinn** [comment]
My gut says this has something to do with how uploads work and the event pool in crt client.  if you look at the gist 14 start up (which is the number of cores on my laptop) and my gut says bc im doing 15 uploads that block eachother, the event loop pool cannot handle that

**Comment by TingDaoK** [comment]
yeah, sorry that I made the mistake to use the default java client instead of CRT client.

From the example I have. I made couple modifications and I made sure I am using the CRT client using the following code:
```
        S3AsyncClient client = S3AsyncClient.crtBuilder().build();
        S3TransferManager mgr = S3TransferManager.builder().s3Client(client).build();
        Path file = Path.of("/Users/dengket/project/graalvm/test-crt/sample-project-crt/file.txt");
        long length = file.toFile().length();
        ArrayList<PipedOutputStream> streams = new ArrayList<PipedOutputStream>();
        ArrayList<InputStream> input_streams = new ArrayList<InputStream>();
        var futures = new ArrayList<CompletableFuture<CompletedUpload>>();

        // Executor service for asynchronous operations
        ExecutorService uploadExecutor = Executors.newFixedThreadPool(22);  // Define thread pool for upload tasks
        ThreadPoolExecutor threadPool = (ThreadPoolExecutor) uploadExecutor;

        for (int i =0; i < 20; i++) {
            var putRequest = PutObjectRequest.builder().bucket("bucket").key("prefix/" + "1").build();
            PipedOutputStream outputStream = new PipedOutputStream();
            streams.add(outputStream);
            InputStream inputStream = new PipedInputStream(outputStream);
            input_streams.add(inputStream);

            var uploadRequest = UploadRequest.builder()
                    .putObjectRequest(putRequest)
                    .requestBody(AsyncRequestBody.fromInputStream(inputStream, length, uploadExecutor))
                    .build();

            futures.add(mgr.upload(uploadRequest).completionFuture());
        }
        int activeCount = threadPool.getActiveCount();
        System.out.println(activeCount);
        try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
            String line;
            while ((line = fis.readLine()) != null) {
                for (var stream : streams) {
                    stream.write((line + "\n").getBytes());
                }
            }

            for (var stream : streams) {
                stream.close();
            }
        }
        activeCount = threadPool.getActiveCount();
        System.out.println(activeCount);

        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
        mgr.close();
        client.close();
        uploadExecutor.shutdown();
```


I did reproduce the hanging, when the number of the threads in the `Executors.newFixedThreadPool()` is too small (in my test above, when I have only 2 threads for the executor, it hangs). 

but, if I get 20 threads, which equals to the number of the requests in concurrent, it didn't hang.

I believe if you have too less threads for the pool, each request will submit a task to the executor. https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/InputStreamWithExecutorAsyncRequestBody.java#L80

and if you have much less threads comparing to the number of requests, it will be blocking each other.

**Comment by LarryFinn** [comment]
@TingDaoK which version of the libraries are you using?  im still seeing it get stuck for some number of concurrent sends regardless of threadpool size (i was originally using a cached threadpool anyway, so that size was moot)

**Comment by LarryFinn** [comment]
I modified the loop to have a little more output (im doing 8 concurrent writes)
```            
            var lineNumber = 0;
            String line;
            while ((line = fis.readLine()) != null) {
                lineNumber++;
                for (int j = 0; j < streams.size(); j++) {
                    LOG.infof("Writing to stream=%d line=%d", j, lineNumber);
                    streams.get(j).write((line + "\n").getBytes());
                }
            }
```
and the output looks like
```

2024-11-04 11:17:24.285  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=1
2024-11-04 11:17:24.288  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=1
2024-11-04 11:17:24.289  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=1
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=2
2024-11-04 11:17:24.290  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=2
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=2
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=3
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=3
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=3
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=3
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=3
2024-11-04 11:17:24.291  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=3
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=3
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=3
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=4
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=4
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=4
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=4
2024-11-04 11:17:24.292  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=4
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=4
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=4
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=4
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=5
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=5
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=5
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=5
2024-11-04 11:17:24.293  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=5
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=5
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=5
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=5
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=6
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=6
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=6
2024-11-04 11:17:24.294  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=6
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=6
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=6
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=6
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=6
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=7
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=7
2024-11-04 11:17:24.295  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=7
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=8
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=8
2024-11-04 11:17:24.296  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=8
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=9
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=9
2024-11-04 11:17:24.297  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=9
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=10
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=10
2024-11-04 11:17:24.298  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=10
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=11
2024-11-04 11:17:24.299  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=2 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=3 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=4 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=5 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=6 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=7 line=11
2024-11-04 11:17:24.300  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=0 line=12
2024-11-04 11:17:25.464  INFO ${sys:PID} [    Test worker] c.a.s.s.u.f.p.LarryTest                  : Writing to stream=1 line=12
```

**Comment by TingDaoK** [comment]
I am using the latest version of crt java, but I also tried to use the one you mentioned.

It could related to your file size, how large is the data you try to upload?

**Comment by LarryFinn** [comment]
code du -h /tmp/TEXT-subscribers-100.csv
 87M       /tmp/TEXT-subscribers-100.csv
code wc -l /tmp/TEXT-subscribers-100.csv
 1000000 /tmp/TEXT-subscribers-100.csv

**Comment by TingDaoK** [comment]
I did reproduce the issue when I bump up the file size. I will reach out to Java SDK team for their support on this.

Quick question, is your use case requiring to use the PipedInputStream and PipedOutputStream to provide data async?

**Comment by LarryFinn** [comment]
One of our big use-cases needs this, or something like this, yes

**Comment by TingDaoK** [comment]
Okay, big thanks for @zoewangg to help out debug this.

Firstly, in your code sample:
```
        try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
            String line;
            while ((line = fis.readLine()) != null) {
                for (var stream : streams) {
                    stream.write((line + "\n").getBytes());
                }
            
```
You are writing into the streams all from the main thread. But, in CRT, we will try to full fill the first request with data. And CRT will keep reading until the data for the first request is enough to be sent as a part.

So, in the main thread, you are writing the same amount of data into all the streams. In the end, CRT client has its own order to work on the requests, and (I believe) Java side blocks on the threads/resource usage as CRT requests the data that was not provided yet, while the main thread is trying to provide the data that CRT is not going to consume and leads to the hang.

Apparently my reproduce code that tries to send the request from another thread pool has a bug, which will lead to multiple threads write into the same stream. Sorry for the confusion.

The proper way to provide data for CRT client is something like:
```
        ExecutorService writeExecutor = Executors.newFixedThreadPool(50);  // Define thread pool for upload tasks

        for (int j = 0; j < streams.size(); j++) {
            final int finalJ = j;
            writeExecutor.submit(() -> {
                    try (BufferedReader fis = new BufferedReader(new FileReader(file.toFile()))) {
                        char[] buffer = new char[chunkSize];
                        int bytesRead;
                        while ((bytesRead = fis.read(buffer)) != -1) {
                            String dataChunk = new String(buffer, 0, bytesRead);
                            logger.info("before writing to stream=" + finalJ + "  bytes=" + dataChunk.length());
                            streams.get(finalJ).write((dataChunk + "\n").getBytes());
                            logger.info("after writing to stream=" + finalJ + "  bytes=" + dataChunk.length());
                        }
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                }
            );

        }

```
So, for each stream, submit a task to read from the file and write to the stream. So that when CRT asks for data, there will be a thread to provide the data and not blocking on others. You can check out some sample code [here](https://github.com/TingDaoK/test-pipestream/commit/d3cbd4e8b34b8926b3281448841719fe4dd5e4e0)

**Comment by LarryFinn** [comment]
hi @TingDaoK thanks for getting back to me!
The problem with the suggested solution is it is functionally different.  in my example im writing different lines of a source file to different s3 files.  so imagine if you have one source file you want to split across 4 destination s3 files, .  in the example you posted it is writing the same file to all different destination s3 files.  is CRT just not capable of doing this?

**Comment by TingDaoK** [comment]
I believe your example here:
```
        try (var fis = new BufferedReader(new FileReader(file.toFile()))) {
            String line;
            while ((line = fis.readLine()) != null) {
                for (var stream : streams) {
                    stream.write((line + "\n").getBytes());
                }
            }
        }
```
is writing the same line to all the streams.

But, to write different lines to different streams, you can still do that by editing how the write works.
eg: if you want to split the file evenly to multiple s3 objects, you can modify the code as following:
```
uploadExecutor.submit(() -> {
                    try (BufferedReader fis = new BufferedReader(new FileReader(file.toFile()))) {
                        char[] buffer = new char[chunkSize];
                        int bytesRead;
                        int lines = 0;
                        while ((bytesRead = fis.read(buffer)) != -1) {
                            lines++;
                            if(lines%streams.size() == finalJ) {
                                String dataChunk = new String(buffer, 0, bytesRead);
                                logger.info("before writing to stream=" + finalJ + "  bytes=" + dataChunk.length());
                                streams.get(finalJ).write((dataChunk + "\n").getBytes());
                                logger.info("after writing to stream=" + finalJ + "  bytes=" + dataChunk.length());
                            }
                        }
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                }
            );
```
Note: you will also need to update your content length carefully.

But anyhow, the idea is the streams should NOT block each other.

**Comment by LarryFinn** [comment]
@TingDaoK oh sorry you're absolutely right about what the example code is doing, this example i gave is a bit different than what we actually do so i got confused.  in your prior comment you wrote
"we have a scheduling logic that will try to get one part for each request and then focus on the first request before working on the others. And CRT will keep reading until the data for the first request is enough to be sent as a part."  does this have anything to do with the aws crt event loop threadpool size or something?  because it seems to be fine for smaller number of threads?  also, how big is the part size typically, or can we configure that?

The reason i am kinda digging into this is because our actual use case is a bit more complicated.  we *could*  buffer some amount of the data in memory to make sure the first concurrent send/request completes its part.  im just a little hesitant about buffering a lot of stuff into memory since we deal with very big datasets.

**Comment by TingDaoK** [comment]
I still don't full understand what resource blocked the program that leads to the hang.

the CRT eventloop pool size is half the CPU numbers. The part size is by default 8 MB, and it's configurable.

I think in the case of the order when you provide the data is different then the order client asks for data, you cannot avoid buffering the data in memory. Either it is buffered before you submit it to the client, or client has to buffer the data somewhere.

In C, we do have an alternative API to provide data asynchronous, which can put some request on hold and work on other requests. It have not integrated with Java yet, but it has integrated with mountpoint, https://github.com/awslabs/mountpoint-s3, which maybe the tool you will be interested in.

If you only want Java support, maybe create a feature request, we can prioritize it. With current API, we cannot change the schedule logic on the client, so that in your case, you will have to buffer the data.

**Comment by TingDaoK** [comment]
you can create multiple clients, but with multiple client, the network eventloop will be increased as each client will try to take the network bandwidth to meet the target throughput.

And with multiple client, the schedule logic will be the same, each client will work on the first request assigned to that client first. 

I guess you can try to have one client to download, and then multiple client to upload, and only one request for each upload client. Also, need to configure the target throughput for each client to meet the total network bandwidth you have.

**Comment by LarryFinn** [comment]
When i do have multiple clients as mentioned there is some weird locking issue if you look at the pictures attached at the first post of this ticket

**Comment by TingDaoK** [comment]
well, in your code sample, you have the same client to work on the upload and download, and the first meta request created is the download, and the client will again work on the first request first. So, it's basically the same thing, the other requests will be blocked.

And you block on the upload in the callback of the download. I believe if your number of request is large enough, it may result in deadlock. 

As I said, you could try to have a client to download, and upload it from other clients, it's not recommended, but maybe a hack around your issue.

**Comment by LarryFinn** [comment]
hey @TingDaoK .  I am saying that's what we do already, we have multiple clients in our production code.  However, it causes this weird locking issue that cause high cpu usage.  I'll add the images again to illustrate.
![375772985-9367274c-f93d-4c4f-b2e5-4dda6d1d6eca](https://github.com/user-attachments/assets/583a772c-918c-482f-b7c3-cda03f7c7acf)

**Comment by TingDaoK** [comment]
> in your code sample, you have the same client to work on the upload and download

Can you provide another code sample that match what you say?

> Sorry it took so long, work got busy and it was hard to make this minimal. I think there are two things at play here:
> 
>     1. we use multiple s3 clients, which isn't a best practice but i wouldnt expect this outcome
> 
>     2. we upload as we download
>        The problem is definitely more related to point 1, but im surprised it happens.  Here's some code, it might not compile exactly because i had some different inputs but i think you can figure it out
> 
> 
> ```
> var s3CrtAsyncClientBuilder =  S3AsyncClient.crtBuilder()
>                 .maxConcurrency(Integer.MAX_VALUE);
>    private void multiClientUpload() {
>         var concurrency = 30;
>         var futures = new ArrayList<CompletableFuture<Integer>>(concurrency);
>         var clients = new ArrayList<S3AsyncClient>(concurrency * 2);
>         var managers = new ArrayList<S3TransferManager>(concurrency * 2);
>         long fileSize = fileSize(SOME_PATH);
>         for (int i = 0; i < concurrency; i++) {
>             var dlClient = s3CrtAsyncClientBuilder.build();
>             clients.add(dlClient);
>             var dlMgr = S3TransferManager.builder().s3Client(dlClient).build();
>             managers.add(dlMgr);
>             var ulClient = s3CrtAsyncClientBuilder.build();
>             clients.add(ulClient);
>             var ulMgr = S3TransferManager.builder().s3Client(ulClient).build();
>             managers.add(ulMgr);
>             futures.add(doRequestAndUpload(SOME_PATH, i, dlMgr, ulMgr, fileSize));
>         }
>         CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
>         managers.forEach(S3TransferManager::close);
>         clients.forEach(S3AsyncClient::close);
>     }
> 
>     private CompletableFuture<Integer> doRequestAndUpload(
>             String path
>             int i,
>             S3TransferManager mgr,
>             S3TransferManager uploadMgr,
>             long fileSize) {
>         return doRequest(path, i, mgr, stream -> {
>             var body =
>                     new InputStreamWithExecutorAsyncRequestBody(AsyncRequestBodyFromInputStreamConfiguration.builder()
>                             .executor(testPool)
>                             .inputStream(stream)
>                             .maxReadLimit(8 * 1024 * 1024)
>                             .contentLength(fileSize)
>                             .build());
>             var uploadRequest = UploadRequest.builder()
>                     .requestBody(body)
>                     .putObjectRequest(PutObjectRequest.builder()
>                             .bucket(BUCKET_HERE)
>                             .key("blarg/" + i + ".csv")
>                             .build())
>                     .build();
>             var result = uploadMgr.upload(uploadRequest).completionFuture().join();
>             return i;
>         });
>     }
> 
>     private long fileSize(String path) {
>         try (var client = s3CrtAsyncClientBuilder.build()) {
>             return client.headObject(b -> b.bucket(BUCKET_HERE).key(path))
>                     .thenApply(HeadObjectResponse::contentLength)
>                     .join();
>         }
>     }
>     private CompletableFuture<Integer> doRequest(
>             String path,
>             int i,
>             S3TransferManager mgr,
>             Function<InputStream, Integer> consumer) {
>         DownloadRequest<ResponseInputStream<GetObjectResponse>> downloadRequest;
>         var downloadRequestBuilder = DownloadRequest.builder()
>                 .getObjectRequest(GetObjectRequest.builder()
>                         .bucket(SOME_BUCKET)
>                         .key(path)
>                         .build());
>             downloadRequest = downloadRequestBuilder
>                     .responseTransformer(AsyncResponseTransformer.toBlockingInputStream())
>                     .build();
>         return mgr.download(downloadRequest)
>                 .completionFuture()
>                 .thenApplyAsync(
>                         input -> {
>                             return consumer.apply(input.result());
>                         },
>                         testPool);
>     }
> ```
> 
> `multiClientUpload` is what triggers the whole thing. I made a cached threadpool as testPool. The file i used is about 500mb with ~6m lines. my guess is the multiple clients need to do some locking JNI wise and that locks across clients

**Comment by TingDaoK** [comment]
sorry, I took another look into your sample code, it's not the same issue I thought. Looks like the multiple client issue is different than what we looked into for the single part.

I am not sure about why, but looks like you use the same thread pool for all the object download and upload, which is one possibility of the cause.

My assumption is:
I "believe" for every download, each 8 MB will trigger a task to deliver the body and added to the thread pool to deliver the body. 
And then you start the upload from the same thread, but use the same thread pool to submit the body for the upload, in which case, the thread pool may already be filled up by all those download parts. 
So, the CRT upload starving waiting for the body to be scheduled in the fully occupied thread pool, and leads to keep looping in the CRT thread and try to get enough data to be sent.
